src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,538 @@
+             * our DockerCloud instances", then we should be careful with it, and do not delete the node (yet).
+             * Even if the containerId originates from a collision, then we would just postpone the deletion
+             * of the node up to that point in time, when that container gets deleted from the DockerContainer
+             * instance (on the next recurrence of this watchdog, we then would detect that situation and the
+             * node gets deleted automatically, as no container with the appropriate identifier could be detected
+             * anymore). 
+             * 
+             * It is to be hoped that collision on container identifiers are happening rarely enough to such that
+             * we do not run into troubles here. Otherwise, still a larger set of broken nodes could pile up...
+             */
+            
+            boolean seenOnDockerInstance = csmMerged.isContainerIdRegistered(dtn.getContainerId());
+            if (seenOnDockerInstance) {
+                // the node and the container still might have a proper mapping => ignore
+                continue;
+            }
+            
+            SlaveComputer computer = dtn.getComputer();
+            if (computer == null) {
+                // Probably the node is being closed down right now, so we shouldn't touch it.
+                continue;
+            }
+            
+            if (!computer.isOffline()) {
+                // the node is still running; we should not touch it.
+                continue;
+            }
+            
+            // the container is already gone for the node, but the node did not notice it yet properly
+            LOGGER.info("{} has container ID {}, but the container does not exist in any docker cloud. Will remove node.", dtn, dtn.getContainerId());
+
+            try {
+                removeNode(dtn);
+                executionStatistics.nodesRemoved++;
+            } catch (IOException e) {
+                LOGGER.warn(String.format("Failed to remove orphaned %s.", dtn), e);
+                executionStatistics.nodesRemovedFailed++;
+            }
+        }
+    }
+
+    private void checkForTimeout(Instant startedTimestamp) {
+        final Instant now = clock.instant();
+        Duration runtime = Duration.between(startedTimestamp, now);
+        
+        if (runtime.compareTo(PROCESSING_TIMEOUT_IN_MS) > 0) {
+            throw new WatchdogProcessingTimeout();
+        }
+    }
+    
+    private static class Statistics {
+        public long executions;
+        public long containersRemovedGracefully;
+        public long containersRemovedGracefullyRuntimeSum;
+        public long containersRemovedForce;
+        public long containersRemovedForceRuntimeSum;
+        public long containersRemovedFailed;
+        public long nodesRemoved;
+        public long nodesRemovedFailed;
+        public long processingTimeout;
+        public long overallRuntime;
```

Comment:
I'm somewhat uneasy about having a lot of directly-accessible fields in a class - that doesn't seem to be very object-oriented and doesn't hide the implementation from the outside.
My _general_ preference is to have classes that hide their fields and only expose methods to the outside.  I accept that there are times when this would result in an impractical amount of methods and hence this _might_ be a situation where following this general preference would result in worse code ... but I'm not 100% sure that this is the case here - it looks a little bit like this was just for convenience.

What are your opinions?

Reply:
Well, I did not look at this case from a "class" coupling perspective, but I more saw this class as a technical substitute of a `struct`. My first idea had been to just declare all the statistical fields as plain fields of the main class. However, as it became more and more, I had the desire to "group" them somehow (they also don't serve a direct purpose for the functionality of that class). 

But given your comment at the bottom, I think it's okay to really refactor it to make the class really a class and not a helper's construct.

Done with 9af913b48aa2693583334a367f3e26f3c7643fee